File: sem1/hw9/9.1/list.cpp
#include "list.h"
#include <stdio.h>

List *createList() {
    return new List{nullptr};
}

void pushList(ListElement *&element, int sity) {
    if (!element) {
        element = new ListElement;
        element->next = nullptr;
        element->town = sity;
        return;
    }
    pushList(element->next, sity);
}

void pushList(List *&list, int sity) {
    pushList(list->element, sity);
}

int findTown(ListElement *element, int number) {

    if (number == -1) {
        return element->town;
    }
    while ((number != element->town) && (element->next != nullptr)) {
        element = element->next;
    }
    if (element->next == nullptr)
        return -1;
    return element->next->town;
}

int findTown(List *list, int number) {
    return findTown(list->element, number);
}

void printList(ListElement *list) {
    while (list) {
        printf("%d ", list->town);
        list = list->next;
    }
}

void printList(List *list) {
    printList(list->element);
}

void deleteOneElement(ListElement *list, int **matrix, int element) {
    while (list) {
        matrix[element][list->town - 1] = -1;
        matrix[list->town - 1][element] = -1;
        list = list->next;
    }
}

void deleteOneElement(List *list, int **matrix, int element) {
    deleteOneElement(list->element, matrix, element);
}

void deleteList(ListElement *element) {
    while (element) {
        ListElement *current = element;
        element = element->next;
        delete current;
    }
}

void deleteList(List *list) {
    deleteList(list->element);
    delete list;
}

File: sem2/hw4/task4_1/ru/liamin/vladimir/List.java
package ru.liamin.vladimir;

/** Implementation of a list */
public class List<T> {

    private int size;
    private ListElement head;

    /**
     * Writing elements to the list
     * @param value value of an element
     * @param index number of the element
     * throws ElementAlreadyExistsExeption exception report having duplicate
     */
    public void add(T value, int index) throws ElementAlreadyExistsExeption {

        if (size == 0) {
            head = new ListElement(value, head);
            size++;
            return;
        }

        ListElement temp = head;

        for (int i = 0; i < index - 2; i++) {
            temp = temp.next;
        }

        size++;
        temp.next = new ListElement(value, temp.next);
        return;
    }

    /**
     * Delete an element in list
     * @param number value of an element
     * @throws ElementDoesNotExistExeption
     */
    public void remove(T number) throws ElementDoesNotExistExeption {
        ListElement temp = head;
        if (!find(number)) {
            throw new ElementDoesNotExistExeption("This item is not there");
        }

        if (head.value == number) {
            head = head.next;
            size--;
            return;
        }

        while ((temp.next != null) && (temp.next.value != number))
            temp = temp.next;

        temp.next = temp.next.next;
        size--;
        return;
    }

    /**
     * Check for emptiness
     * @return return true if list is empty
     */
    public boolean isEmpty() {
        return head == null;
    }

    /**
     * Count of elements in list
     * @return number of elements in list
     */
    public int count() {
        return size;
    }

    /** Print list element */
    public void printList() {
        System.out.println("Elements of list: ");
        ListElement current = head;
        while (current != null) {
            System.out.println(current.value);
            current = current.next;
        }
    }

    /**
     * Looks for an item in the list
     * @param number value of an element
     * @return true if an element in list
     */
    public boolean find(T number) {
        ListElement temp = head;
        for (int i = 0; i < size; i++) {
            if (temp.value == number)
                return true;
            temp = temp.next;
        }
        return false;
    }

    /** Clear list */
    public void clear() {
        size = 0;
        head = null;

    }

    private class ListElement<T> {
        private T value;
        private ListElement next;

        public ListElement(T number, ListElement next) {
            value = number;
            this.next = next;
        }
    }

}